tutorials/tensordictmodule.ipynb

```diff
@@ -0,0 +1,2042 @@
+    "        super().__init__(\n",
+    "            AttentionBlockTensorDict(to_name, from_name, to_dim, to_len, from_dim, latent_dim, num_heads),\n",
+    "            TensorDictModule(FFN(to_dim, 4*to_dim), in_keys=[to_name], out_keys=[\"X_out\"]),\n",
+    "            TensorDictModule(SkipLayerNorm(to_len, to_dim), in_keys=[to_name, \"X_out\"], out_keys=[to_name]),\n",
+    "        )\n",
+    "class TransformerBlockDecoderTensorDict(TensorDictSequence):\n",
+    "    def __init__(self, to_name, from_name, to_dim, to_len, from_dim, latent_dim, num_heads):\n",
+    "        super().__init__(\n",
+    "            AttentionBlockTensorDict(to_name, to_name, to_dim, to_len, to_dim, latent_dim, num_heads),\n",
+    "            TransformerBlockEncoderTensorDict(to_name, from_name, to_dim, to_len, from_dim, latent_dim, num_heads)\n",
+    "        )"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 38,
+   "id": "eba41ed9",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "TensorDict(\n",
+       "    fields={\n",
+       "        Attn: Tensor(torch.Size([8, 2, 3, 10]), dtype=torch.float32),\n",
+       "        K: Tensor(torch.Size([8, 2, 10, 5]), dtype=torch.float32),\n",
+       "        Q: Tensor(torch.Size([8, 2, 3, 5]), dtype=torch.float32),\n",
+       "        V: Tensor(torch.Size([8, 2, 10, 5]), dtype=torch.float32),\n",
+       "        X_from: Tensor(torch.Size([8, 10, 6]), dtype=torch.float32),\n",
+       "        X_out: Tensor(torch.Size([8, 3, 5]), dtype=torch.float32),\n",
+       "        X_to: Tensor(torch.Size([8, 3, 5]), dtype=torch.float32)},\n",
+       "    batch_size=torch.Size([8]),\n",
+       "    device=cpu,\n",
+       "    is_shared=False)"
+      ]
+     },
+     "execution_count": 38,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "to_dim = 5\n",
+    "from_dim = 6\n",
+    "latent_dim = 10\n",
+    "to_len = 3\n",
+    "from_len = 10\n",
+    "batch_size = 8\n",
+    "num_heads = 2\n",
+    "\n",
+    "tokens = TensorDict(\n",
+    "    {\n",
+    "        \"X_to\": torch.randn(batch_size, to_len, to_dim),\n",
+    "        \"X_from\": torch.randn(batch_size, from_len, from_dim)\n",
+    "    },\n",
+    "    batch_size=[batch_size]\n",
+    ")\n",
+    "\n",
+    "transformer_block = AttentionBlockTensorDict(\n",
+    "    \"X_to\",\n",
+    "    \"X_from\",\n",
+    "    to_dim,\n",
+    "    to_len,\n",
+    "    from_dim,\n",
+    "    latent_dim,\n",
+    "    num_heads\n",
+    ")\n",
+    "\n",
+    "transformer_block(tokens)\n",
+    "\n",
+    "tokens"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a89c3eea",
+   "metadata": {},
+   "source": [
+    "The output of the transformer layer can now be found at tokens[\"X_to\"]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 39,
+   "id": "cb06a532",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
```

Comment:
printing data like this is a bit cumbersome. I would only print stuff that is "human-readable".